File: dotnet/CucumberExpressions.Tests/CombinatorialGeneratedExpressionFactoryTest.cs
using System;
using System.Collections.Generic;
using CucumberExpressions.Generation;
using FluentAssertions;
using Xunit;

namespace CucumberExpressions.Tests;

public class CombinatorialGeneratedExpressionFactoryTest : CucumberExpressionTestBase
{
    private const string WORD = "\\w+";

    [Fact]
    public void Generates_multiple_expressions()
    {
        var first = new List<IParameterType>();
        first.Add(new StubParameterType<Color>("color", WORD));
        first.Add(new StubParameterType<CssColor>("csscolor", WORD));

        var second = new List<IParameterType>();
        second.Add(new StubParameterType<Date>("date", WORD));
        second.Add(new StubParameterType<DateTime>("datetime", WORD));
        second.Add(new StubParameterType<TimeSpan>("timestamp", WORD));
        var parameterTypeCombinations = new List<List<IParameterType>>()
        {
            first, second
        };

        var factory = new CombinatorialGeneratedExpressionFactory(
                "I bought a {{{0}}} ball on {{{1}}}",
                parameterTypeCombinations
        );
        var generatedExpressions = factory.GenerateExpressions();
        var expressions = new List<string>();
        foreach (var generatedExpression in generatedExpressions)
        {
            var source = generatedExpression.GetSource();
            expressions.Add(source);
        }
        expressions.Should().BeEquivalentTo(new[] {
                "I bought a {color} ball on {date}",
                "I bought a {color} ball on {datetime}",
                "I bought a {color} ball on {timestamp}",
                "I bought a {csscolor} ball on {date}",
                "I bought a {csscolor} ball on {datetime}",
                "I bought a {csscolor} ball on {timestamp}"
        });
    }

    public class Color
    {
    }

    public class CssColor
    {
    }

    public class Date
    {
    }
}

File: dart/test/combinatorial_generated_expression_factory_test.dart
import 'package:cucumber_expressions/src/combinatorial_generated_expression_factory.dart';
import 'package:cucumber_expressions/src/parameter_type.dart';
import 'package:test/test.dart';

void main() {
  group('CombinatorialGeneratedExpressionFactory', () {
    test('generates multiple expressions', () {
      final parameterTypeCombinations = <List<ParameterType<Object?>>>[
        [
          ParameterType<String?>(
              'color', RegExp('red|blue|yellow'), null, (s) => s.first,
              useForSnippets: false, preferForRegexpMatch: true),
          ParameterType<String?>(
              'csscolor', RegExp('red|blue|yellow'), null, (s) => s.first,
              useForSnippets: false, preferForRegexpMatch: true),
        ],
        [
          ParameterType<String?>(
              'date', RegExp(r'\d{4}-\d{2}-\d{2}'), null, (s) => s.first,
              useForSnippets: false, preferForRegexpMatch: true),
          ParameterType<String?>(
              'datetime', RegExp(r'\d{4}-\d{2}-\d{2}'), null, (s) => s.first,
              useForSnippets: false, preferForRegexpMatch: true),
          ParameterType<String?>(
              'timestamp', RegExp(r'\d{4}-\d{2}-\d{2}'), null, (s) => s.first,
              useForSnippets: false, preferForRegexpMatch: true),
        ],
      ];

      final factory = CombinatorialGeneratedExpressionFactory(
        'I bought a {{0}} ball on {{1}}',
        parameterTypeCombinations,
      );
      final expressions =
          factory.generateExpressions().map((ge) => ge.source).toList();
      expect(
        expressions,
        equals([
          'I bought a {color} ball on {date}',
          'I bought a {color} ball on {datetime}',
          'I bought a {color} ball on {timestamp}',
          'I bought a {csscolor} ball on {date}',
          'I bought a {csscolor} ball on {datetime}',
          'I bought a {csscolor} ball on {timestamp}',
        ]),
      );
    });
  });
}
